### PR TITLE
UHM-4832 Migration of treatment obs from HIV EMR to OpenMRS

### PR DESCRIPTION
**Pull request by brandones**
https://pihemr.atlassian.net/browse/UHM-4832

This only accounts for "20. Treatment / Phrophylactic measures" through "22. Treatment anti-TB"

index

Documented https://pihemr.atlassian.net/wiki/spaces/ZL/pages/1052770307/ZL+HIV+Meds+Migration+Plan

ObsLoading takes ~10 minutes

### CODE REVIEW

**Review by cioan — APPROVED** [review]
Thanks @brandones ! LGTM!

**Review by mogoodrich — APPROVED** [review]
Generally LGTM, though I didn't review the actual migration queries in any detail.  It seems like the best way for reviewing all those would be the task of creating example data of all permutations in the source HIV TEST system, doing them migration, and then reviewing in the destination OpenMRS instance. A very tedious task, but probably necessary?

Also would be good to get @mseaton 's high level review before merging as he is closest to this.

File: etl/src/main/java/org/pih/hivmigration/etl/sql/TreatmentObsMigrator.groovy [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+        executeMysql("Load intermediate prophylaxis table", '''
+            DROP PROCEDURE IF EXISTS load_prophylaxis;
+            DELIMITER $$ ;
+            CREATE PROCEDURE load_prophylaxis(_txName varchar(80))
+            BEGIN
+                INSERT INTO hivmigration_prophylaxis
+                    (name, start, cont, stopped, other_value, source_encounter_id, encounter_date)
+                SELECT
+                    _txName,
+                    IF(MAX(base.ordered = _txName), TRUE, FALSE) AS start,
```
Comment (mseaton) [inline etl/src/main/java/org/pih/hivmigration/etl/sql/TreatmentObsMigrator.groovy:83]: Can you explain what these "MAX" functions are doing?  Doesn't seem right.

File: etl/src/main/java/org/pih/hivmigration/etl/sql/TreatmentObsMigrator.groovy [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+                    other.comments AS other_value,
+                    base.source_encounter_id,
+                    he.encounter_date
+                FROM hivmigration_ordered_other base
+                JOIN (
+                    SELECT source_encounter_id, comments
+                    FROM hivmigration_ordered_other
+                    WHERE ordered = 'other_prophylaxis') other ON base.source_encounter_id = other.source_encounter_id
+                JOIN hivmigration_encounters he ON base.source_encounter_id = he.source_encounter_id
+                WHERE base.ordered IN (_txName, CONCAT(_txName, '_continue'))
```
Comment (mseaton) [inline etl/src/main/java/org/pih/hivmigration/etl/sql/TreatmentObsMigrator.groovy:95]: What is this where clause for that checks on _continue?

File: etl/src/main/java/org/pih/hivmigration/etl/sql/TreatmentObsMigrator.groovy [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+            FROM hivmigration_arv_regimen;
+
+            INSERT INTO tmp_obs (source_encounter_id, concept_uuid, value_coded_uuid, obs_group_id)
+            SELECT
+                source_encounter_id,
+                concept_uuid_from_mapping('CIEL', '1282'),
+                CASE comments
+                    WHEN 'azt_3tc_efv' THEN concept_uuid_from_mapping('CIEL', '160124')
+                    WHEN 'azt_3tc_nvp' THEN concept_uuid_from_mapping('CIEL', '1652')
+                    ELSE concept_uuid_from_mapping('CIEL', '5424')  -- other
```
Comment (mseaton) [inline etl/src/main/java/org/pih/hivmigration/etl/sql/TreatmentObsMigrator.groovy:240]: This doesn't seem right.  We have lots more coded values.  We should map them, even if they are not on the current form.

**Review by mseaton — COMMENTED** [review]
Really hard to comprehensively review.  We need some better actual test cases, running against source and target, that can allow us to validate a bit more.

**Comment by mogoodrich** [comment]
Good point @mseaton ... would be interesting to see if we could create some test data (or "harvest" some good test cases from teh HIV EMR) and create some actual test cases around this?

File: etl/src/main/java/org/pih/hivmigration/etl/sql/TreatmentObsMigrator.groovy (hunk 1 above)
Comment (brandones) [reply etl/src/main/java/org/pih/hivmigration/etl/sql/TreatmentObsMigrator.groovy:83]: Yeah, this is a pattern for checking whether any member of the group matches the condition. I guess `TRUE > FALSE`.

File: etl/src/main/java/org/pih/hivmigration/etl/sql/TreatmentObsMigrator.groovy (hunk 2 above)
Comment (brandones) [reply etl/src/main/java/org/pih/hivmigration/etl/sql/TreatmentObsMigrator.groovy:95]: It says "create a row in the new prophylaxis form if either 'initée' or 'continuée' are checked in the old form"

File: etl/src/main/java/org/pih/hivmigration/etl/sql/TreatmentObsMigrator.groovy (hunk 3 above)
Comment (brandones) [reply etl/src/main/java/org/pih/hivmigration/etl/sql/TreatmentObsMigrator.groovy:242]: Ok, I added the other codes that exist in non-trivial quantities

Screenshot from 2020-09-29 13-48-27

**Comment by brandones** [comment]
Holding off until I get iterative obs loading working...

**Comment by brandones** [comment]
Waiting to merge until iterative obs migration is in: https://github.com/PIH/hivmigration/pull/40

**Pull request merged**